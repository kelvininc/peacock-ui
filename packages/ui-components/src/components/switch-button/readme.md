# *<kv-switch-button>*

<!-- Auto Generated Below -->


## Usage

### Angular / javascript

```html
<!-- Default -->
<kv-switch-button></kv-switch-button>

<!-- Labeled -->
<kv-switch-button label="Switch"></kv-switch-button>

<!-- Disabled -->
<kv-switch-button disabled></kv-switch-button>

<!-- ON/OFF -->
<kv-switch-button state="ON"></kv-switch-button>
<kv-switch-button state="OFF"></kv-switch-button>
```


### React

```tsx
import React from 'react';

import { KvSwitchButton } from '@kelvininc/react-ui-components';

export const SwitchButtonExample: React.FC = () => (
  <>
    {/*-- Default --*/}
    <KvSwitchButton/>

    {/*-- Labeled --*/}
	<KvSwitchButton label="Switch"/>

	{/*-- Disabled --*/}
	<KvSwitchButton disabled/>

	{/*-- ON/OFF --*/}
	<KvSwitchButton state="ON"/>
	<KvSwitchButton state="OFF"/>
  </>
);

```


### Stencil

```tsx
import { Component, h } from '@stencil/core';

@Component({
  tag: 'switch-button-example',
  styleUrl: 'switch-button-example.css',
  shadow: true,
})
export class SwichButtonExample {
  render() {
    return [
      	// Default
		<kv-switch-button></kv-switch-button>

		// Labeled
		<kv-switch-button label="Switch"></kv-switch-button>

		// Disabled
		<kv-switch-button disabled></kv-switch-button>

		// ON/OFF
		<kv-switch-button state="ON"></kv-switch-button>
		<kv-switch-button state="OFF"></kv-switch-button>
    ];
  }
}
```



## Properties

| Property   | Attribute  | Description                                 | Type                                              | Default                  |
| ---------- | ---------- | ------------------------------------------- | ------------------------------------------------- | ------------------------ |
| `disabled` | `disabled` | (optional) If `true` the button is disabled | `boolean`                                         | `false`                  |
| `label`    | `label`    | (optional) Button's label                   | `string`                                          | `''`                     |
| `state`    | `state`    | (optional) If `ON` the button is ON         | `ESwitchButtonState.OFF \| ESwitchButtonState.ON` | `ESwitchButtonState.OFF` |


## Events

| Event               | Description                         | Type                                                           |
| ------------------- | ----------------------------------- | -------------------------------------------------------------- |
| `switchStateChange` | Emitted when switch's state changes | `CustomEvent<ESwitchButtonState.OFF \| ESwitchButtonState.ON>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
