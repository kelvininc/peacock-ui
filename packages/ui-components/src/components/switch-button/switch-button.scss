@import '../../assets/styles/globals';

$btn-height: 28px;
$btn-width: 56px;
$btn-radius: 4px;
$sqr-size: 20px;
$sqr-radius: 2px;
$icon-size: 16px;

.switch-button-container {
	display: inline-flex;
	align-items: center;

	user-select: none;
}

.label {
	@include kv-font-label-small-uppercase-semibold;
	flex-shrink: 0;
	color: kv-color('neutral-4');
	padding-right: $spacing-3x;
}

.switch-button {
	box-sizing: border-box;

	height: $btn-height;
	width: $btn-width;

	cursor: pointer;
	border-radius: $btn-radius;
	padding: $spacing;

	background-color: kv-color('neutral-2');
	transition: background-color 0.3s ease-out;

	&--disabled {
		cursor: default;
		pointer-events: none;
		opacity: 0.5;
	}

	&--on {
		background-color: kv-color(success);

		.check-square {
			transform: translateX(28px);

			.svg-icon {
				fill: kv-color(success);
			}
		}
	}
}

.check-square {
	display: inline-flex;
	align-items: center;
	justify-content: center;

	width: $sqr-size;
	height: $sqr-size;
	transition: transform 0.3s ease-out;
	border-radius: $sqr-radius;
	background-color: kv-color(success, contrast);

	font-size: 0;

	.svg-icon {
		fill: kv-color('neutral-2');
		width: $icon-size;
		height: $icon-size;
	}
}
