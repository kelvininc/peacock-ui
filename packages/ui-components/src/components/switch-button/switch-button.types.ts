import { EventEmitter } from '@stencil/core';
import { EComponentSize } from '../../types';

export interface ISwitchButton {
	/** (optional) If `true` the button is disabled. Default `false` */
	disabled: boolean;
	/** (optional) If `true` the button is ON. Default `false` */
	checked: boolean;
	/** (optional) Button's size. Default `EComponentSize.Large` */
	size: EComponentSize;
}

export interface ISwitchButtonEvents {
	/** Emitted when switch's state changes */
	switchChange: EventEmitter<boolean>;
}
