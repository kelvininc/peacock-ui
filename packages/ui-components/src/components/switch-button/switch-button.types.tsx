export enum ESwitchButtonState {
	ON = 'on',
	OFF = 'off'
}

export enum ESwitchButtonSize {
	Small = 'sm',
	Large = 'lg'
}
