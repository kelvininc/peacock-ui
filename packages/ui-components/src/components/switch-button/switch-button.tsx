import { Component, Event, EventEmitter, Host, Prop, State, Watch, h } from '@stencil/core';
import { ESwitchButtonSize, ESwitchButtonState } from './switch-button.types';
import throttle from 'lodash/throttle';

/**
 * @part icon-svg - The switch icon.
 * @part icon-square - The switch icon square container.
 * @part button - The switch button.
 */
@Component({
	tag: 'kv-switch-button',
	styleUrl: 'switch-button.scss',
	shadow: true
})
export class KvSwitchButton {
	/** (optional) Button's label */
	@Prop() label: string = '';

	/** Watch `label` property for changes and update `hasLabel` accordingly */
	@Watch('label')
	labelHandler(newValue?: string) {
		this.hasLabel = newValue != null && newValue !== '';
	}

	/** (optional) If `true` the button is disabled */
	@Prop({ reflect: true }) disabled: boolean = false;

	/** Watch `disabled` property for changes and update `isDisabled` accordingly */
	@Watch('disabled')
	disabledHandler(newValue: boolean) {
		this.isDisabled = newValue === true;
	}

	/** (optional) If `ON` the button is ON */
	@Prop({ reflect: true, mutable: true }) state: ESwitchButtonState = ESwitchButtonState.OFF;
	/** (optional) Button's size */
	@Prop() size: ESwitchButtonSize = ESwitchButtonSize.Large;

	/** Watch `state` property for changes and update `isOn` accordingly */
	@Watch('state')
	stateHandler(newValue: ESwitchButtonState) {
		this.isOn = newValue === ESwitchButtonState.ON;
	}

	/** Whether the label exist and it's not empty */
	@State() hasLabel: boolean = this.label != null && this.label !== '';
	/** Whether the state is ON or `true` */
	@State() isOn: boolean = this.state === ESwitchButtonState.ON;
	/** Whether the state is ON or `true` */
	@State() isDisabled: boolean = this.disabled === true;

	/** Emitted when switch's state changes */
	@Event() switchStateChange: EventEmitter<ESwitchButtonState>;

	private onSwitchClick: () => void;

	private onStateChange() {
		if (this.isDisabled) {
			return;
		}

		this.state = this.isOn ? ESwitchButtonState.OFF : ESwitchButtonState.ON;
		this.switchStateChange.emit(this.state);
	}

	connectedCallback() {
		this.onSwitchClick = throttle(() => this.onStateChange(), 300);
	}

	render() {
		const iconName = this.isDisabled ? 'kv-lock' : 'kv-done-all';

		return (
			<Host>
				<div class="switch-button-container">
					{this.hasLabel && <span class="label">{this.label}</span>}
					<div class="switch-button-content">
						<div
							class={{
								'switch-button': true,
								'switch-button--disabled': this.isDisabled,
								'switch-button--on': this.isOn,
								'switch-button--lg': this.size === ESwitchButtonSize.Large,
								'switch-button--sm': this.size === ESwitchButtonSize.Small
							}}
							part="button"
							onClick={this.onSwitchClick}
						>
							<div class="icon-square" part="icon-square">
								<kv-svg-icon name={iconName} part="icon-svg" />
							</div>
						</div>
					</div>
				</div>
			</Host>
		);
	}
}
