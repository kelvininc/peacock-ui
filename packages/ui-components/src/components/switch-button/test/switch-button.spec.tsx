import { SpecPage } from '@stencil/core/internal';
import { KvSwitchButton } from '../switch-button';
import { newSpecPage } from '@stencil/core/testing';

describe('Switch Button (unit tests)', () => {
	let page: SpecPage;
	let component: KvSwitchButton;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSwitchButton],
				html: `<kv-switch-button></kv-switch-button>`,
			});
			component = page.rootInstance;
		})

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize `isOn` with false', () => {
			expect(component.isOn).toBe(false)
		});

		it('should initialize `hasLabel` with false', () => {
			expect(component.hasLabel).toBe(false)
		});
	})

	describe('when has a label', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSwitchButton],
				html: `<kv-switch-button label="Switch"></kv-switch-button>`,
			});
			component = page.rootInstance;
		})

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize `hasLabel` with true', () => {
			expect(component.hasLabel).toBe(true)
		});
	})

	describe('when is disabled', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSwitchButton],
				html: `<kv-switch-button disabled></kv-switch-button>`,
			});
			component = page.rootInstance;
		})

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when is ON', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSwitchButton],
				html: `<kv-switch-button state="on"></kv-switch-button>`,
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize `isOn` with true', () => {
			expect(component.isOn).toBe(true)
		});
	});
});
