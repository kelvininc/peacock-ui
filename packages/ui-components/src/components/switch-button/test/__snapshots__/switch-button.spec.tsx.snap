// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Button (unit tests) when has a label should match the snapshot 1`] = `
<kv-switch-button label="Switch" state="off">
  <mock:shadow-root>
    <div class="switch-button-container">
      <span class="label">
        Switch
      </span>
      <div class="switch-button-content">
        <div class="switch-button switch-button--lg" part="button">
          <div class="icon-square" part="icon-square">
            <kv-svg-icon name="kv-done-all" part="icon-svg"></kv-svg-icon>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</kv-switch-button>
`;

exports[`Switch Button (unit tests) when is ON should match the snapshot 1`] = `
<kv-switch-button state="on">
  <mock:shadow-root>
    <div class="switch-button-container">
      <div class="switch-button-content">
        <div class="switch-button switch-button--lg switch-button--on" part="button">
          <div class="icon-square" part="icon-square">
            <kv-svg-icon name="kv-done-all" part="icon-svg"></kv-svg-icon>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</kv-switch-button>
`;

exports[`Switch Button (unit tests) when is disabled should match the snapshot 1`] = `
<kv-switch-button disabled="" state="off">
  <mock:shadow-root>
    <div class="switch-button-container">
      <div class="switch-button-content">
        <div class="switch-button switch-button--disabled switch-button--lg" part="button">
          <div class="icon-square" part="icon-square">
            <kv-svg-icon name="kv-lock" part="icon-svg"></kv-svg-icon>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</kv-switch-button>
`;

exports[`Switch Button (unit tests) when uses default props should match the snapshot 1`] = `
<kv-switch-button state="off">
  <mock:shadow-root>
    <div class="switch-button-container">
      <div class="switch-button-content">
        <div class="switch-button switch-button--lg" part="button">
          <div class="icon-square" part="icon-square">
            <kv-svg-icon name="kv-done-all" part="icon-svg"></kv-svg-icon>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</kv-switch-button>
`;
